entrypoint/src/mill/entrypoint/MillBootstrapModule.scala

```diff
@@ -0,0 +1,214 @@
+    }
+    def scalaVersion = "2.13.10"
+
+    def parseBuildFiles = T.input {
+      MillBootstrapModule.parseBuildFiles(base)
+    }
+
+    def ivyDeps = T {
+      Agg.from(parseBuildFiles()._2.map(mill.scalalib.Dep.parse(_))) ++
+      Seq(ivy"com.lihaoyi::mill-moduledefs:${Versions.millModuledefsVersion}")
+    }
+
+    def scriptSources = T.sources {
+      for ((p, s) <- parseBuildFiles()._1) yield PathRef(p)
+    }
+
+    def generatedSources = T {
+      val (scripts, ivy, importTrees, errors) = parseBuildFiles()
+      if (errors.nonEmpty) Result.Failure(errors.mkString("\n"))
+      else Result.Success(
+        for ((p, (_, s)) <- scriptSources().zip(parseBuildFiles()._1)) yield {
+          val relative = p.path.relativeTo(base)
+          val segments = Seq.fill(relative.ups)("^") ++ relative.segments
+          val dest = T.dest / segments
+
+          os.write(
+            dest,
+            MillBootstrapModule.top(relative, p.path / os.up, segments.dropRight(1), p.path.baseName) +
+            s +
+            MillBootstrapModule.bottom,
+            createFolders = true
+          )
+          PathRef(dest)
+        }
+      )
+    }
+
+    def scriptImportGraph = T {
+      val (scripts, ivy, scriptImportGraph, errors) = parseBuildFiles()
+      scriptImportGraph
+    }
+
+    override def allSourceFiles: T[Seq[PathRef]] = T {
+      Lib.findSourceFiles(allSources(), Seq("scala", "java", "sc")).map(PathRef(_))
+    }
+
+    def unmanagedClasspath = mill.define.Target.input {
+      mill.api.Loose.Agg.from(enclosingClasspath.map(p => mill.api.PathRef(p))) ++
+      lineNumberPluginClasspath()
+    }
+
+    def scalacPluginIvyDeps = Agg(
+      ivy"com.lihaoyi:::scalac-mill-moduledefs-plugin:${Versions.millModuledefsVersion}"
+    )
+
+    def scalacPluginClasspath = super.scalacPluginClasspath() ++ lineNumberPluginClasspath()
+
+    def lineNumberPluginClasspath: T[Agg[PathRef]] = T {
+      mill.modules.Util.millProjectModule(
+        "MILL_LINENUMBERS",
+        "mill-entrypoint-linenumbers",
+        repositoriesTask()
+      )
+    }
+  }
+}
+
+object MillBootstrapModule{
+  def top(relative: os.RelPath, base: os.Path, pkg: Seq[String], name: String) = {
+    val foreign =
+      if (pkg.nonEmpty || name != "build") {
+        // Computing a path in "out" that uniquely reflects the location
+        // of the foreign module relatively to the current build.
+
+        // Encoding the number of `/..`
+        val ups = if (relative.ups > 0) Seq(s"up-${relative.ups}") else Seq()
+        val segs =
+          Seq("foreign-modules") ++
+          ups ++
+          relative.segments.init ++
+          Seq(relative.segments.last.stripSuffix(".sc"))
+
+        val segsList = segs.map(pprint.Util.literalize(_)).mkString(", ")
+        s"Some(_root_.mill.define.Segments.labels($segsList))"
+      } else "None"
+    s"""
+       |package millbuild${pkg.map("." + _).mkString}
+       |import _root_.mill._
```

Comment:
Created a discussion to remove this magic import:
https://github.com/com-lihaoyi/mill/discussions/2381

Reply:
I'm for it, but we should probably try to not make this PR any larger. Esp. when it comes to document the changes, having more but smaller PRs helps to document changed and update documentation.